### PR TITLE
Stardew Valley: Fix a bug where locations in logic would disappear from universal tracker as items get sent

### PR DESCRIPTION
**Pull request by Jouramie**
## What is this fixing or adding?
This bug was found when a player reported that museumsanity stuff would disappear from UT. It was caused by the incrementation of the `total_progression_item` in the `create_item` method of the world. UT call `create_item` when an item is collected, which would increase the `total_progression_item`. 

This value is used by some rules so that locations that are deemed grindy or are only available after multiple in game months or years are moved to later spheres, so they are not expected from the player right away. Increasing the `total_progression_item` would make those locations unavailable, even tho the amount of collected progression item would increase as well. 

For instance, imagine a world with 100 progression items where 4 are already collected. In these settings, locations that expect 4% of the items to be collected should be available, but they would not before that fix, since the total progression items would also have been increase by for. 4/104 != 4%. 

Another bug was also fixed where Golden Walnuts would count as multiple items when calculating the current progress in items, because they are added as events in the `collect` method of the world without increasing the `total_progression_items`.

I also removed the `prevent_sweep=True/False` from stardew tests, since none of them were necessary and someone was copy-pasting them all around for no reason.

## How was this tested?
Ran the tests.

I also generated a world, collected the necessary items to make the museumsanity locations available. Sent tons of Stardrop with the cheat console and made sure that locations were still marked as in logic.

## If this makes graphical changes, please attach screenshots.
N/A

### CODE REVIEW

File: worlds/stardew_valley/__init__.py [hunk 1]
```diff
@@ -171,15 +170,22 @@ def create_items(self):
...
 
         setup_early_items(self.multiworld, self.options, self.player, self.random)
         self.setup_player_events()
         self.setup_victory()
 
+        # This is really a best-effort to get the total progression items count. The actual total can be impacted by the start_inventory_from_pool when items
+        # are removed from the bool, and probably when plando is applied as well... But since this is only used to spread grindy items across spheres, it's not
+        # a big deal if it's not 100% accurate.
+        self.total_progression_items += sum(1 for i in self.multiworld.precollected_items[self.player] if i.advancement)
+        self.total_progression_items += sum(1 for i in self.multiworld.get_filled_locations(1) if i.advancement)
```
Comment (Exempt-Medic) [inline worlds/stardew_valley/__init__.py:185]: This takes a `player` arg, so passing in `1` seems very weird

Comment (Jouramie) [reply worlds/stardew_valley/__init__.py:185]: Good catch, this should be the player id

File: worlds/stardew_valley/__init__.py [hunk 2]
```diff
@@ -171,15 +170,22 @@ def create_items(self):
...
 
         self.multiworld.itempool += created_items
 
         setup_early_items(self.multiworld, self.options, self.player, self.random)
         self.setup_player_events()
         self.setup_victory()
 
+        # This is really a best-effort to get the total progression items count. The actual total can be impacted by the start_inventory_from_pool when items
+        # are removed from the bool, and probably when plando is applied as well... But since this is only used to spread grindy items across spheres, it's not
+        # a big deal if it's not 100% accurate.
```
Comment (Exempt-Medic) [inline worlds/stardew_valley/__init__.py:183]: Three things:
1. `bool` -> `pool`
2. FWIW `start_inventory_from_pool` *is* being accounted for here, since they're made pre-collected long before here and Stardew's `get_filler_item_name` doesn't create additional progression items.
3. This is not only used for spreading items around, both the allsanity and perfection goals use it. This *is* a bug, and creates unwinnable seeds. For example: Plando'ing Blueberry Seeds onto Blueberry Seeds Shipsanity while also adding additional progression items through plando in a `minimal` world will result in a logically beatable, physically unbeatable world. This bug isn't really relevant to this code though as it already exists in Stardew

File: worlds/stardew_valley/__init__.py (hunk 1 above)
Comment (Exempt-Medic) [reply worlds/stardew_valley/__init__.py:185]: Realized I could just make a suggestion
```suggestion
        self.total_progression_items += sum(1 for i in self.multiworld.get_filled_locations(self.player) if i.advancement)
```

File: worlds/stardew_valley/__init__.py (hunk 2 above)
Comment (Jouramie) [reply worlds/stardew_valley/__init__.py:183]: > FWIW start_inventory_from_pool is being accounted for here, since they're made pre-collected long before here and Stardew's get_filler_item_name doesn't create additional progression items.

My understanding from reading Main.py is that they will be counted two times in the total progression items. They are pre-collected between `generate_early` and `create_regions`, but a duplicated version of the item will also be created in `create_items`. The total will not be decreased when the items are removed from the item pool. 

I might be wrong, tho. Would not be the first time I misunderstand what's happening in core.

Comment (Exempt-Medic) [reply worlds/stardew_valley/__init__.py:183]: Oh then that's *also* a bug and isn't being accounted for and creates unwinnable seeds

Comment (Jouramie) [reply worlds/stardew_valley/__init__.py:183]: You're right about the plando bug, it can create unwinnable seeds. 

The simplest solution for both the plando and start inv issues seems to refresh the count of progression items in the `pre_fill` hook, but I'm not sure changing the total progression items mid-generation won't create other issues. It will need to be in another pr so it can be tested more thoroughly.

Comment (Jouramie) [reply worlds/stardew_valley/__init__.py:183]: A short term solution could be to block generation of Allsanity and Perfection goals with minimal accessibility.

File: worlds/stardew_valley/stardew_rule/state.py [hunk 3]
```diff
@@ -102,16 +105,19 @@ def value(self):
...
             return True
 
         total_count = 0
         for item, item_count in player_state.items():
+            if item in stardew_world.excluded_from_total_progression_items:
+                continue
+
             total_count += item_count
             if total_count >= needed_count:
                 return True
```
Comment (Exempt-Medic) [inline worlds/stardew_valley/stardew_rule/state.py:123]: I messages this in DMs, but any thoughts on this kind of thing:
```py
        player_state = {item: item_count for item, item_count in state.prog_items[self.player].items() if item not in stardew_world.excluded_from_total_progression_items}

        if needed_count <= len(player_state):
            return True

        if needed_count <= sum(player_state.values()):
            return True
```

File: worlds/stardew_valley/__init__.py (hunk 2 above)
Comment (Exempt-Medic) [reply worlds/stardew_valley/__init__.py:183]:
```suggestion
        # This is really a best-effort to get the total progression items count. The actual total can be impacted by the start_inventory_from_pool when items
        # are removed from the pool, and probably when plando is applied as well... But since this is only used to spread grindy items across spheres, it's not
        # a big deal if it's not 100% accurate.
```

**Review by Exempt-Medic — APPROVED** [review]
Overall the changes LGTM. Two small comments and the bug I found while investigating these changes is best left solved another time. Tested some random generations which all worked and the logic looked to be intact still.

File: worlds/stardew_valley/__init__.py (hunk 2 above)
Comment (Exempt-Medic) [reply worlds/stardew_valley/__init__.py:183]: > A short term solution could be to prevent generation of Allsanity and Perfection goals with minimal accessibility. All progression items need to be collected to win, so it does not make that much sense to generate with minimal accessibility.

That's such a good hotfix, wow. I love it

File: worlds/stardew_valley/stardew_rule/state.py (hunk 3 above)
Comment (Jouramie) [reply worlds/stardew_valley/stardew_rule/state.py:123]: I'll benchmark it

**Review by agilbert1412 — APPROVED** [review]
Approving the code, I talked about it with Jouramie and Medic. the mentioned bugs have another related but not same PR.
I also ran the tests for what it's worth

File: worlds/stardew_valley/__init__.py (hunk 2 above)
Comment (Jouramie) [reply worlds/stardew_valley/__init__.py]: Nvm that start inventory from pool bug, it does not exist. Stardew does not support the option. We made the default start inventory option automatically remove items from the pool. Removed items are replaced by resource packs and accounted for properly because they are removed in the `create_items` method before the progression item count.

**Review by agilbert1412 — APPROVED** [review]

File: worlds/stardew_valley/__init__.py [hunk 4]
```diff
@@ -120,18 +120,32 @@ def force_change_options_if_incompatible(self):
...
             logger.warning(
                 f"Goal '{goal_name}' requires Ginger Island. Exclude Ginger Island setting forced to 'False' for player {self.player} ({player_name})")
+
         if exclude_ginger_island and self.options.walnutsanity != Walnutsanity.preset_none:
             self.options.walnutsanity.value = Walnutsanity.preset_none
             player_name = self.multiworld.player_name[self.player]
             logger.warning(
                 f"Walnutsanity requires Ginger Island. Ginger Island was excluded from {self.player} ({player_name})'s world, so walnutsanity was force disabled")
 
+        if goal_is_perfection and self.options.accessibility != Accessibility.option_full:
```
Comment (Exempt-Medic) [inline worlds/stardew_valley/__init__.py:137]: So `items` would be fine here (it ensures all progression items reachable) but Stardew also doesn't support `items` so "not full" technically works. You could make it `== minimal` but 🤷

File: worlds/stardew_valley/__init__.py [hunk 5]
```diff
@@ -121,18 +120,32 @@ def force_change_options_if_incompatible(self):
...
+            self.options.accessibility.value = Accessibility.option_full
+            player_name = self.multiworld.player_name[self.player]
```
Comment (Exempt-Medic) [inline worlds/stardew_valley/__init__.py:139]: You can use this now
```suggestion
            player_name = self.player_name
```

Comment (Exempt-Medic) [inline worlds/stardew_valley/__init__.py:139]: You can use this now:
```suggestion
            player_name = self.player_name
```

File: worlds/stardew_valley/__init__.py [hunk 6]
```diff
@@ -121,18 +120,32 @@ def force_change_options_if_incompatible(self):
...
+            logger.warning(
+                f"Goal 'Perfection' requires full accessibility. Accessibility setting forced to 'Full' for player {self.player} ({player_name})")
+
+        if self.options.goal == Goal.option_allsanity and self.options.accessibility != Accessibility.option_full:
```
Comment (Exempt-Medic) [inline worlds/stardew_valley/__init__.py:143]: You can't have two goals, so an `elif` should be fine
```suggestion
        elif self.options.goal == Goal.option_allsanity and self.options.accessibility != Accessibility.option_full:
```

File: worlds/stardew_valley/__init__.py [hunk 7]
```diff
@@ -121,18 +120,32 @@ def force_change_options_if_incompatible(self):
...
 
+        if goal_is_perfection and self.options.accessibility != Accessibility.option_full:
+            self.options.accessibility.value = Accessibility.option_full
+            player_name = self.player_name
+            logger.warning(
+                f"Goal 'Perfection' requires full accessibility. Accessibility setting forced to 'Full' for player {self.player} ({player_name})")
+
+        elif self.options.goal == Goal.option_allsanity and self.options.accessibility != Accessibility.option_full:
+            self.options.accessibility.value = Accessibility.option_full
+            player_name = self.multiworld.player_name[self.player]
```
Comment (Exempt-Medic) [inline worlds/stardew_valley/__init__.py:145]:
```suggestion
            player_name = self.player_name
```

**Review by Exempt-Medic — APPROVED** [review]
Reaffirming approval

**Comment by Jouramie** [comment]
@Berserker66 Ready when you are!

**Review by Berserker66 — APPROVED** [review]
